Implements secret handling

matt-bernhardt opened the pull request:
### Why are these changes being introduced:

* We need a way to handle secret values like credentials and API keys.
* Two specific plugins which need this are our SMTP plugin, and the Sentry integration.

### Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/engx-204

### How does this address that need:

This adds a few things to the application to help with secret handling:

* It adds a code block to wp-config.php to read the file created by the Terminus Secrets Plugin (which needs to be installed by each engineer). For now at least, each value in the secrets file needs to be handled explicitly rather than using a generic parsing.
* It adds a must-use plugin which defines a dashboard widget to list which secret values have been defined in that environment. This is a supplement to the Terminus Secrets Plugin, and is only visible to network admins.
* It adds the SMTP plugin, which will be configured as part of this PR. (The Sentry plugin was added, but not configured, in a prior PR).
* The required and optional secrets are defined in the project readme.

### Document any side effects to this change:

* Through this PR we'll get a working Sentry and SMTP integration.
* This PR also restores the upstream `readme.md` file in the repository root, and moves our local version into the `.github` directory.

## Developer

### Secrets

- [x] All new secrets have been added to Pantheon tiers
- [x] Relevant secrets have been updated in Github Actions
- [x] All new secrets documented in README

### Documentation

- [x] Project documentation has been updated
- [ ] No documentation changes are needed

### Accessibility

- [ ] ANDI or Wave has been run in accordance to
      [our guide](https://mitlibraries.github.io/guides/basics/a11y.html) and
      all issues introduced by these changes have been resolved or opened as new
      issues (link to those issues in the Pull Request details above)

### Stakeholder approval

- [ ] Stakeholder approval has been confirmed
- [x] Stakeholder approval is not needed

### Dependencies

YES dependencies are updated


## Code Reviewer

- [x] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [x] The changes have been verified
- [x] The documentation has been updated or is unnecessary
- [x] New dependencies are appropriate or there were no changes

matt-bernhardt committed:
Implements secret handling

** Why are these changes being introduced:

* We need a way to handle secret values like credentials and API keys.
* Two specific plugins which need this are our SMTP plugin, and the
  Sentry integration (at least so far).

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/engx-204

** How does this address that need:

This adds a few things to the application to help with secret handling:

* It adds a code block to wp-config.php to read the file created by the
  Terminus Secrets Plugin (which needs to be installed by each engineer)
* It adds a must-use plugin which defines a dashboard widget to list
  which secret values have been defined in that environment. This is a
  supplement to the Terminus Secrets Plugin, and is only visible to
  network admins.
* It adds the SMTP plugin, which will be configured as part of this PR.
  (The Sentry plugin was added, but not configured, in a prior PR).
* It adds information about these secrets to our project readme.

** Document any side effects to this change:

* Through this PR we'll get a working Sentry and SMTP integration.

matt-bernhardt committed:
Move our readme content to the .github directory

** Why are these changes being introduced:

* We want to display our fork's readme content here, but leave the
  upstream readme in place in the project root.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/engx-204

** How does this address that need:

* This moves our existing readme into the .github directory, and
  restores the readme in the repo root from the upstream fork.
* Subsequent commits in this branch will add documentation about our
  secrets.

** Document any side effects to this change:

* This _is_ the side effect in this PR.

JPrevost commented on web/wp-config.php:
```diff
@@ -50,7 +50,27 @@
+    $secrets = json_decode( file_get_contents( $secrets_file ), 1 );
+
+    // WP Mail SMTP configuration is required.
+    define( 'WPMS_ON', true );
```
Do we want this forced on here? Scenario: there may be some aspects with secrets we want to load locally (even if the secrets themselves are different than in pantheon). I can imagine we wouldn't want to enable SMTP locally even if there are other config we _do_ want to enable that would follow this block of code.

JPrevost replied:
Nevermind, I read the notes on the plugin page and this just turns on the use of constants and isn't the toggle for mail on/off.

JPrevost commented on web/app/mu-plugins/mitlib-secrets-widget/mitlib-secrets-widget.php:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @package MITlib Secrets Widget
+ * @author Matt Bernhardt
+ * @link https://github.com/mitlibraries/mitlib-secrets-widget
```
I'm curious how (or if) we'll want to represent this type of information as we start collapsing to a mono-repo.

matt-bernhardt replied:
We've not followed a consistent method for attributing authorship to local plugins, IIRC. I think Hattie and Philip both have their name on plugins they wrote, while I've tried to use just "MIT Libraries" - or list both the org and person.

Whatever we want to do, I think we should do it consistently, but this has been pretty far down the priority list of things to address thus far.

JPrevost replied:
Yeah. I agree with the inconsistent attribution as a "we can talk about it later" thing, but in this case I was referring to the `@link` which seems a bit wrong unless we intend to keep the old repo around longterm. This can _also_ be a later thing.

matt-bernhardt replied:
Oh, yeah. Hmm. I've made https://mitlibraries.atlassian.net/browse/ENGX-212 to figure out an approach for this sort of metadata within a monorepo. I suspect that `@link` may not be a required field, so it could be better to just not use that field here.

JPrevost commented on .github/README.md:
```diff
@@ -0,0 +1,185 @@
+### Required Environment variables
+
+- `WPMS_SMTP_PASS` Password associated with the username in `WPMS_SMTP_USER`.
+- `WPMS_SMTP_USER` Username expected by the email server to send emails from WordPress. Associated with `WPMS_SMTP_PASS`.
```
I'm curious if it would be clearer to understand where things are configured if we moved all SMTP configuration to secrets even though portions are not truly "secret". Thoughts?

JPrevost replied:
Never mind, I think those to values are supported by the plugin and the rest is expected to be handled via the plugin page. This is fine as-is. I'm leaving the previous comment so you can see where my questions were when thinking about this code but I don't want to change how the plugin works as it won't be worth it.

JPrevost commented on .github/README.md:
```diff
@@ -0,0 +1,185 @@
+
+## Required environment and secrets
+
+We use the [Terminus Secrets Plugin](https://github.com/pantheon-systems/terminus-secrets-plugin) to manage environment variables within Pantheon. Please avoid using the phpdotenv library and `.env` files until [ENGX-209](https://mitlibraries.atlassian.net/browse/ENGX-209) is closed.
```
Adding a note like:

`Please see the Terminus Secrets Plugin readme for installation and usage instructions.` might be helpful here.

matt-bernhardt replied:
Sure thing - I can push up a commit to adjust this.

JPrevost reviewed (approved):
I've left some comments that I figured out myself in place to demonstrate some misunderstanding/confusion I was having along the way. They don't require changes at all.

It may be helpful to explicitly link to the Installation/Use of the secrets plugin but that is minor and optional as I followed the link myself and read how to install/use it with the instructions as-is.

matt-bernhardt committed:
Address code review feedback

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/engx-204

** How does this address that need:

* This updates the wording in the readme around the secrets plugin,
  clarifying where installation and usage information can be found.
